File: examples/Iterators.java
// --------------
// Iterators.java
// --------------

import java.util.ArrayList;
import java.util.Iterator;

final class MyArrayList<T> {
    private ArrayList<T> _x;

    private final class MyIterator implements Iterator<T> {
        Iterator<T> _i;

        public MyIterator () {
            _i = _x.iterator();}

        public boolean hasNext () {
            return _i.hasNext();}

        public T next () {
            return _i.next();}

        public void remove () {
            _i.remove();}}

    public MyArrayList () {
        _x = new ArrayList<T>();}

    public void add (T v) {
        _x.add(v);}

    public Iterator<T> iterator () {
        return new MyIterator();}}

final class Iterators {
    public static void main (String[] args) {
        System.out.println("Iterators.java");

        {
        ArrayList<Integer> x = new ArrayList<Integer>();
        x.add(2);
        x.add(3);
        Iterator<Integer> p = x.iterator();
        assert  p.hasNext();
        assert  p.next() == 2;
        assert  p.hasNext();
        assert  p.next() == 3;
        assert !p.hasNext();
        }

        {
        MyArrayList<Integer> x = new MyArrayList<Integer>();
        x.add(2);
        x.add(3);
        Iterator<Integer> p = x.iterator();
        assert  p.hasNext();
        assert  p.next() == 2;
        assert  p.hasNext();
        assert  p.next() == 3;
        assert !p.hasNext();
        }

        System.out.println("Done.");}}

File: examples/Iterators.py
#!/usr/bin/env python

# ------------
# Iterators.py
# ------------

print "Iterators.py"

class List1 (object) :
    class Iterator (object) :
        def __init__ (self, y) :
            self.i = iter(y.x)

        def __iter__ (self) :
            return self

        def next (self) :
            try :
                return self.i.next()
            except StopIteration :
                raise

    def __init__ (self) :
        self.x = []

    def __getitem__ (self, i) :
        return self.x[i]

    def __iter__ (self) :
        return List1.Iterator(self)

    def __len__ (self) :
        return len(self.x)

    def append (self, v) :
        self.x.append(v)

class List2 (object) :
    def __init__ (self) :
        self.x = []

    def __getitem__ (self, i) :
        return self.x[i]

    def __iter__ (self) :
        for v in self.x :
            yield v

    def __len__ (self) :
        return len(self.x)

    def append (self, v) :
        self.x.append(v)

class List3 (object) :
    def __init__ (self) :
        self.x = []

    def __getitem__ (self, i) :
        return self.x[i]

    def __iter__ (self) :
        return (v for v in self.x)

    def __len__ (self) :
        return len(self.x)

    def append (self, v) :
        self.x.append(v)

def test (T) :
    x = T()
    x.append(2)
    x.append(3)
    assert hasattr(x, "__len__")
    assert x.__len__()      == 2
    assert len(x)           == 2
    assert hasattr(x, "__getitem__")
    assert x.__getitem__(1) == 3
    assert x[1]             == 3
    assert hasattr(x, "__iter__")
    i = x.__iter__()
    i = iter(x)
    assert i is iter(i)
    assert i.next() == 2
    assert i.next() == 3
    try :
        assert i.next() == 4
        assert False
    except StopIteration :
        pass

test(list)
test(List1)
test(List2)
test(List3)

print "Done."
